packages/hawtio/src/plugins/diagnostics/flight-recorder-service.ts

```diff
@@ -0,0 +1,260 @@
+import { ILogger, Logger } from '@hawtiosrc/core'
+import { jolokiaService, MBeanNode, workspace } from '../shared'
+
+export interface IFlightRecorderService {
+  getFlightRecoderMBean(): Promise<MBeanNode | undefined>
+  setUp(): Promise<any>
+  startRecording(userJfrSettings?: UserJfrSettings): Promise<any>
+  stopRecording(): Promise<any>
+  downloadRecording(id: number, name: string): Promise<any>
+}
+
+export enum RecordingState {
+  NOT_CREATED,
+  CREATED,
+  RECORDING,
+  STOPPED,
+}
+
+export interface UserJfrSettings {
+  limitType: 'duration' | 'maxSize' | 'unlimited'
+  limitValue: number
+  recordingNumber: number
+  dumpOnExit: boolean
+  name: string
+  configuration: string
+}
+
+export type Recording = {
+  number: string
+  size: string
+  file: string
+  time: any
+  canDownload: boolean
+  downloadLink: string
+}
+
+export type CurrentRecording = {
+  state: RecordingState
+  number?: number
+}
+
+export type JfrConfig = {
+  name: string
+  label: string
+  description: string
+}
+
+class FlightRecorderService implements IFlightRecorderService {
+  private jfrLogger: ILogger = Logger.get('jfr-service')
+
+  public jfrMBean?: MBeanNode
+  public jfrConfigs?: JfrConfig[]
+  public userJfrSettings?: UserJfrSettings
+
+  public recordings: Array<Recording> = []
+  public currentRecording: CurrentRecording = { state: RecordingState.NOT_CREATED }
+  public initialized: boolean = false
+
+  async setUp(): Promise<FlightRecorderService> {
+    if (!this.jfrMBean) {
+      const jfr = await this.getFlightRecoderMBean()
+
+      if (!jfr) return this
+    }
+
+    await this.retrieveConfigurations()
+    await this.retrieveRecordings()
+    await this.retrieveSettings()
+
+    this.initialized = true
+
+    return this
+  }
```

Comment:
I believe it should be within the `getFlightRecorder()` method and not be invoked from the client of the service explicitly.
```
public jfrMBean?: MBeanNode
```
should then be
```
private jfrMBean?: Promise<MBeanNode | null>
```
and you can handle the three cases: 1) not initialised yet, 2) initialised and the mbean available, 3) initialiased already but the mbean not available